scripts/esbuild/mock-doc.ts

```diff
@@ -0,0 +1,73 @@
+import type { BuildOptions as ESBuildOptions } from 'esbuild';
+import fs from 'fs-extra';
+import { join } from 'path';
+
+import { writeSizzleBundle } from '../bundles/plugins/sizzle-plugin';
+import { bundleDts } from '../utils/bundle-dts';
+import { BuildOptions } from '../utils/options';
+import { writePkgJson } from '../utils/write-pkg-json';
+import { getBaseEsbuildOptions, getEsbuildAliases, runBuilds } from './util';
+
+export async function buildMockDoc(opts: BuildOptions) {
+  const inputDir = join(opts.buildDir, 'mock-doc');
+  const srcDir = join(opts.srcDir, 'mock-doc');
+  const outputDir = opts.output.mockDocDir;
+
+  // bundle d.ts
+  await bundleMockDocDts(opts, inputDir, outputDir);
+
+  writePkgJson(opts, outputDir, {
+    name: '@stencil/core/mock-doc',
+    description: 'Mock window, document and DOM outside of a browser environment.',
+    main: 'index.cjs',
+    module: 'index.js',
+    types: 'index.d.ts',
+    sideEffects: false,
+  });
+
+  const mockDocAliases = getEsbuildAliases();
+
+  const sizzlePath = await writeSizzleBundle(opts);
+  mockDocAliases['sizzle'] = sizzlePath;
+
+  const mockDocBuildOptions: ESBuildOptions = {
+    ...getBaseEsbuildOptions(),
+    entryPoints: [join(srcDir, 'index.ts')],
+    bundle: true,
+    alias: mockDocAliases,
+    logLevel: 'info',
+    target: 'node16',
+  };
+
+  const esmOptions: ESBuildOptions = {
+    ...mockDocBuildOptions,
+    format: 'esm',
+    outfile: join(outputDir, 'index.js'),
+  };
+
+  const cjsOptions: ESBuildOptions = {
+    ...mockDocBuildOptions,
+    format: 'cjs',
```

Comment:
Do we no longer need the [`CJS_INTRO` and `CJS_OUTRO` portions](https://github.com/ionic-team/stencil/blob/9d2d0a0612caafb88c559907c6dec1a39e760615/scripts/bundles/mock-doc.ts#L41-L42) of the build?

Reply:
I don't believe its necessary

Reply:
Can you expound on why?

Reply:
as far as I can tell the `CJS_INTRO` and `CJS_OUTRO` snippets are here to allow the cjs bundle of mock doc to be used outside of a node.js / commonjs context. In the Rollup build the whole bundle is wrapped with

```js
var mockDoc = (function(exports) {
'use strict';
```

and then

```js
if (typeof module !== "undefined" && module.exports) {
  module.exports = exports;
}
return exports;
})({});
```

so basically the whole thing gets wrapped in an IIFE which will override the `exports` identifier within the scope of the IIFE and then if `typeof module !== "undefined"` (i.e. if we're in a node environment) only then it will set `module.exports` to be `exports`. I don't think this is necessary anymore since we don't support in-browser compilation. Also afaict Esbuild doesn't output cjs code using the [`exports` shortcut](https://nodejs.org/api/modules.html#exports-shortcut) but instead always references `module.exports`, so I think that including this intro and outro as-is would actually just cause `module.exports` to be set to `{}`.

Instead Esbuild, in one place, sets `module.exports` like this:

```js
module.exports = __toCommonJS(mock_doc_exports);
```

the `__toCommonJs` is a helper emitted by Esbuild that produces a commonJS module from an object with getters on it - you can check it out locally to see how it works (kinda big to paste here haha)

but anyway, all that is to say that I don't think the IIFE wrapping does anything we need anymore. I've also been able to build mock-doc using Esbuild and then run tests in a sample project without issues.